core/src/main/scala/org/bykn/bosatsu/pattern/SeqPattern.scala

```diff
@@ -369,63 +367,54 @@ object SeqPattern {
           case (Cat(_: SeqPart1[A], _), Empty) =>
             // Cat(SeqPart1[A], _) does not match Empty
             p1 :: Nil
-          case (_, Cat(Wildcard, _)) if p2.matchesAny =>
-            // matches anything
+          case (_, _) if p2.matchesAny || subset(p1, p2) =>
+            // p2 has to be bigger
             Nil
-          case (_, _) if subset(p1, p2) => Nil
           case (Cat(Wildcard, t1@Cat(Wildcard, _)), _) =>
             // unnormalized
             difference(t1, p2)
           case (_, Cat(Wildcard, t2@Cat(Wildcard, _))) =>
             // unnormalized
             difference(p1, t2)
-          case (Cat(Wildcard, Cat(a1, t1)), _) if isAny(a1) =>
+          case (Cat(Wildcard, Cat(a1: SeqPart1[A], t1)), _) if isAny(a1) =>
             // *. == .*, push Wildcards to the end
             difference(Cat(a1, Cat(Wildcard, t1)), p2)
-          case (_, Cat(Wildcard, Cat(a2, t2))) if isAny(a2) =>
+          case (_, Cat(Wildcard, Cat(a2: SeqPart1[A], t2))) if isAny(a2) =>
             // *. == .*, push Wildcards to the end
             difference(p1, Cat(AnyElem, Cat(Wildcard, t2)))
-          case (_, _) if intersection(p1, p2).isEmpty => p1 :: Nil
           case (Cat(Wildcard, t1), Empty) =>
             if (!t1.matchesEmpty) p1 :: Nil
             else {
               // *:t1 = t1 + _:p1
               // _:p1 - [] = _:p1
               unifyUnion(Cat(AnyElem, p1) :: difference(t1, Empty))
             }
+          case (_, _) if disjoint(p1, p2) => p1 :: Nil
           case (Cat(h1: SeqPart1[A], t1), Cat(Wildcard, t2)) =>
-            // h1:t1 - (*:t2) = h1:t1 - t2 - _:p2
-            val d12 = {
-              val dtail = difference(t1, p2)
-              if (dtail == (t1 :: Nil)) p1 :: Nil
```

Comment:
you can't use difference this way, it is an upper bound, you have to directly use disjointness.